### DISCUSSION TITLE
Playground: Show visualization for the relationship between different Types Definitions across Types

### DISCUSSION THREAD

**Discussion by KamalAman** [openfga/community#299] [Ideas]
### Checklist

- [X] I agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md).

### Describe the problem you'd like to have solved

Right now it is difficult to visualize how a specific Type Definition is compose of direct and implied relationships. 

### Describe the ideal solution

Ideally there would be some way to visual the implied relationship / the complete relationship hierarchy for a given Type Definitions. 

The example would be the `Repo:Admin -> Org:Repo Admin`
![Screenshot 2024-04-15 at 9 10 40 AM](https://github.com/openfga/openfga/assets/2692357/8843cb50-4b94-4498-ae7e-61fc85985ac2)

I don't think just showing all relationships N-degrees away makes sense since it would be chaotic. When you hover a Type, then show the Types full definition with direct and implied relationships

### Alternatives and current workarounds

Currently, I don't think there is any easy way way to see the implied relationships. You have to to read the model and assertions and reason about it / draw it out.

### Additional context

As a model gets more complex, it is hard to reason about the implied relationships. Having visualizations for implied definitions would definitely help being simplicity back

**Comment by KamalAman** [comment]
This maybe related to https://github.com/orgs/openfga/discussions/122

**Comment by rhamzeh** [comment]
Thanks for the suggestion @KamalAman - yes, we recognize there's a lot of deficiencies in our types previewer. Our current visualization predates the `1.1` schema, and is currently on our back-burner. We would like to have it fixed, but it's low in priority

What would help us here is collecting ideas for what you imagine this would look like.

As you said, for any production-grade model, showing all the links would make the graph unreadable, and there's a question on how to show the links.

Please feel free to add your suggestions if you have them, anything would help, and it might spark further conversation from the community

**Comment by KamalAman** [comment]
Yeah, only show direct relationships by default. 
When ever you hover a node, you show faint connections for all implied relationships N-degrees away.
If you click on node, you show prominent connections for all implied relationships N-degrees away.

I believe that should provide the user all of the context without overloading them. 

Use case: 

If I hover/click on `repo:writer`, i should be able to see that `org:admins` have a implied reference to be a writer through the path of  `repo:maintainer -> repo:admin -> org:admins`

**Reply by rhamzeh** [reply]
Thanks for the feedback!

Would you also expect to see a link to the types that are directly assignable?

e.g.

```
type repo
  relations
    define reader: [user, user:*, team#member, org#member]
```

Do you expect to see a link to the user type, to the member relation on the team type and to the member relation on the org type?

How do you imagine `and/but not` or complex relations being drawn?

**Reply by KamalAman** [reply]
For directly assignable, probably show them in a less  prominent way.

As fas for `and/but not` probably use some sort of `broken/termination` line.
I think for complex relations (I assume you're talking about conditions), just show them as implied relationships. With maybe some sort of indicator on them. 

For all of these different conditions, I think it is important to show it N'degress out to provide the full context when you interact with a node.

**Comment by zalsader** [comment]
What would be helpful in reducing clutter is to have the option of only showing the relationships only when you click on a certain relation.
